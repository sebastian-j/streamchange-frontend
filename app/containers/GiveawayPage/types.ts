import { ActionType } from 'typesafe-actions';
import * as actions from './actions';

interface IBan {
  channelId: string,
  description: string,
  endsAt: string,
}

interface IStream {
  ownerId: string;
  thumbnailUrl: string;
  title: string;
  videoId: string;
}

type Ban = IBan;
type Stream = IStream;

/* --- STATE --- */

interface GiveawayState {
  readonly authKey: string;
  readonly ban: Ban | null;
  readonly stream: Stream;
}

/* --- ACTIONS --- */
type AppActions = ActionType<typeof actions>;

/* --- EXPORTS --- */

type ContainerState = GiveawayState;
type ContainerActions = AppActions;

export { ContainerState, ContainerActions, Ban, Stream };
