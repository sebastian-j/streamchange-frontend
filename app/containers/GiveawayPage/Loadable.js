/**
 * Asynchronously loads the component for GiveawayPage
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
