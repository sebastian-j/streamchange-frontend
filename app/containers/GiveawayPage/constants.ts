/*
 *
 * GiveawayPage constants
 *
 */
enum ActionTypes {
  CHANGE_STREAM_PROPERTIES = 'app/GiveawayPage/CHANGE_STREAM_PROPERTIES',
  CHANGE_BAN_STATUS = 'app/GiveawayPage/CHANGE_BAN_STATUS',
  LOAD_AUTH_KEY = 'app/GiveawayPage/LOAD_AUTH_KEY',
  SEND_TELEMETRY_DATA = 'app/GiveawayPage/SEND_TELEMETRY_DATA',
}

export default ActionTypes;
