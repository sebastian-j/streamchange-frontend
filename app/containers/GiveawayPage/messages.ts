/*
 * GiveawayPage Messages
 *
 * This contains all the text for the GiveawayPage container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.containers.GiveawayPage';

export default defineMessages({
  pageTitle: {
    id: `${scope}.title`,
    defaultMessage: 'StreamChange Giveaway',
  },
  historyLink: {
    id: `${scope}.HistoryWidget.ButtonText`,
    defaultMessage: 'Win history',
  },
  historyEmptyTooltip: {
    id: `${scope}.HistoryWidget.EmptyTooltip`,
    defaultMessage:
      'Winner history is stored here. History is currently empty.',
  },
  historyWidgetTableTitle: {
    id: `${scope}.HistoryWidget.TableTitle`,
    defaultMessage: 'Last winners',
  },
  historyWidgetWarningText: {
    id: `${scope}.HistoryWidget.WarningText`,
    defaultMessage:
      'Storing too long history can affect app performance. Clear winner history when you no longer need it.',
  },
  leaveStreamBtn: {
    id: `${scope}.leaveStreamBtn`,
    defaultMessage: 'Leave stream',
  },
});
