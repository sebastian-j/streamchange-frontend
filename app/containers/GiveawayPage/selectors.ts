import { createSelector } from 'reselect';
import { ApplicationRootState } from 'types';
import { initialState } from './reducer';

/**
 * Direct selector to the GiveawayPage state domain
 */
const selectGiveawayPage = (state: ApplicationRootState) =>
  state.giveawayPage || initialState;

const makeSelectAuthKey = () =>
  createSelector(selectGiveawayPage, (state) => state.authKey);

const makeSelectBanStatus = () =>
  createSelector(selectGiveawayPage, (state) => state.ban);

const makeSelectStreamInfo = () =>
  createSelector(selectGiveawayPage, (state) => state.stream);

export {
  selectGiveawayPage,
  makeSelectAuthKey,
  makeSelectBanStatus,
  makeSelectStreamInfo,
};
