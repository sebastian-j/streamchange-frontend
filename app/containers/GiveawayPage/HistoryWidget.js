import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { NavLink } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { withStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';

import messages from './messages';
import db from '../../components/YoutubeWorker/db';

const StyledLink = styled(NavLink)`
  border: none;
  border-radius: 6px;
  color: ${(props) => props.theme.buttonTextColor};
  font-size: 1.05rem;
  height: 70%;
  padding: 5px 8px;
  position: relative;
  margin: 0 15px 0 0;
  text-decoration: none;
  &:hover {
    svg.border-hover {
      stroke-dashoffset: 0;
    }
  }
  .border {
    position: absolute;
    top: 0;
    bottom: 0;
    right: 0;
    width: 100%;
    height: 100%;
    margin: auto;
  }
  .border-initial {
    border-radius: 6px;
    border: 1px solid;
    opacity: 0.2;
  }
  svg.border-hover {
    stroke: currentColor;
    stroke-dasharray: 1;
    stroke-dashoffset: 1;
    transition: stroke-dashoffset .5s cubic-bezier(.22,.28,.36,1);
    rect {
      transform-origin: 50% 50%;
      vector-effect: non-scaling-stroke;
    }
  }
  @media (orientation: portrait) {
    line-height: 3em;
  }
`;

const WinnerTableTitle = styled.div`
  font-size: 14px;
  margin: 5px 0;
  text-align: center;
`;

const WarningContent = styled.div`
  font-size: 13px;
`;
const Td = styled.td`
  padding-right: 5px;
  vertical-align: middle;
`;

const ExtendedTooltip = withStyles((theme) => ({
  tooltip: {
    backgroundColor: 'rgba(225,246,246,0.9)',
    color: 'rgba(0, 0, 0, 0.87)',
    maxWidth: '300px',
    fontSize: theme.typography.pxToRem(12),
    border: '1px solid #949499',
  },
}))(Tooltip);

const HistoryWidget = () => {
  const [winners, setWinners] = useState([]);
  const [warning, setWarning] = useState(false);
  const getHistory = () => {
    db.table('history')
      .reverse()
      .toArray()
      .then((items) => {
        setWarning(items.length > 100);
        const it = Array.from(items);
        if (it.length > 3) it.length = 3;
        setWinners(it);
      })
      .catch(() => {
        setWinners([]);
      });
  };

  useEffect(() => {
    getHistory();
  }, []);

  return (
    <ExtendedTooltip
      title={
        <>
          {!warning && winners.length === 0 && (
            <WarningContent>
              <FormattedMessage {...messages.historyEmptyTooltip} />
            </WarningContent>
          )}
          {warning && (
            <WarningContent>
              <FormattedMessage {...messages.historyWidgetWarningText} />
            </WarningContent>
          )}
          {winners.length > 0 && (
            <WinnerTableTitle>
              <FormattedMessage {...messages.historyWidgetTableTitle} />
            </WinnerTableTitle>
          )}
          <table>
            <tbody>
              {winners.map((item) => (
                <tr key={item.createdAt}>
                  <Td>
                    <img src={item.imageUrl} alt="Logo" width="32px" />
                  </Td>
                  <Td>{item.displayName}</Td>
                  <Td>{item.prize}</Td>
                </tr>
              ))}
            </tbody>
          </table>
        </>
      }
    >
      <StyledLink onMouseEnter={getHistory} to="/giveaway-history">
        <span className="border border-initial" />
        <svg className="border border-hover" fill="none">
          <rect width="100%" height="100%" rx="6px" pathLength="1"/>
        </svg>
        <FormattedMessage {...messages.historyLink} />
        {warning && (
          <span role="img" aria-label="warning">
              ⚠️
          </span>
        )}
      </StyledLink>
    </ExtendedTooltip>
  );
};

export default HistoryWidget;
