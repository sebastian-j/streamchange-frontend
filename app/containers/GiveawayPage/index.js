import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import styled from 'styled-components';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import { FormattedMessage } from 'react-intl';

import messages from './messages';
import { makeSelectBanStatus, makeSelectStreamInfo } from './selectors';
import { changeStreamProperties, sendTelemetryData } from './actions';
import { useInjectSaga } from 'utils/injectSaga';
import saga from './saga';
import HistoryWidget from './HistoryWidget';
import WelcomeDialog from '../../components/WelcomeDialog';
import YoutubeWorker from '../../components/YoutubeWorker';
import SettingsDialog from '../../components/SettingsDialog';
import SupportInformation from '../../components/SupportInformation';
import { API_KEY } from '../../config';

const TopBar = styled.div`
  background-color: ${(props) => props.theme.panelBackground};
  display: flex;
  justify-content: space-between;
  @media (orientation: portrait) {
    flex-direction: column;
  }
`;

const StreamInfo = styled.div`
  height: 5vh;
`;

const StreamImg = styled.img`
  height: 100%;
`;

const StreamTitle = styled.span`
  color: ${(props) => props.theme.staticTextColor};
  margin-left: 10px;
`;

const TopButtons = styled.div`
  align-items: center;
  display: flex;
  @media (orientation: portrait) {
    display: flex;
    justify-content: space-between;
    margin: 30px 10px 4px 10px;
  }
`;

const StyledButton = styled(Button)`
  span {
    color: ${(props) => props.theme.color};
  }
`;

const GiveawayPage = (props) => {
  const [error, setError] = useState(null);
  useInjectSaga({ key: 'giveawayPage', saga: saga });

  const leaveStream = () => {
    const streamProps = {
      ownerId: '',
      thumbnailUrl: '',
      title: '',
      videoId: '',
    };
    props.changeStreamProperties(streamProps);
    sessionStorage.removeItem('gv-videoId');
    sessionStorage.removeItem('gv-title');
    sessionStorage.removeItem('gv-thumbnailUrl');
    window.location.reload();
  };

  const launchWorker = (vidId) => {
    axios
      .get(
        `https://www.googleapis.com/youtube/v3/videos?part=snippet%2C+liveStreamingDetails&id=${vidId}&key=${API_KEY}`,
      )
      .then((res) => {
        if (res.data.items.length === 0) {
          setError('notVideo');
        } else if (res.data.items[0].snippet.liveBroadcastContent === 'none') {
          setError('notStream');
        } else {
          const stream = res.data.items[0];
          const streamProps = {
            ownerId: stream.snippet.channelId,
            thumbnailUrl: stream.snippet.thumbnails.medium.url,
            title: stream.snippet.title,
            videoId: vidId,
          };
          props.changeStreamProperties(streamProps);
          props.sendTelemetryData(streamProps);
          sessionStorage.setItem('gv-videoId', vidId);
          sessionStorage.setItem('gv-title', streamProps.title);
          sessionStorage.setItem('gv-thumbnailUrl', streamProps.thumbnailUrl);
          sessionStorage.setItem('gv-ownerId', streamProps.ownerId);
        }
      })
      .catch((err) => {
        if (err.response && err.response.data && err.response.data.error) {
          if (err.response.data.error.errors[0].reason.includes('Exceeded')) {
            setError('quotaExceeded');
          }
        } else {
          const streamProps = {
            ownerId: '',
            thumbnailUrl:
              'https://i.ytimg.com/vi/HwsGz6csNA0/maxresdefault.jpg',
            title: 'TytuÅ‚ nieznany',
            videoId: vidId,
          };
          props.changeStreamProperties(streamProps);
          sessionStorage.setItem('gv-videoId', vidId);
        }
      });
  };

  const receiveVideo = (videoLink) => {
    if (videoLink.includes('v=')) {
      const vidId = videoLink.split('v=')[1].split('&')[0].split('/')[0];
      launchWorker(vidId);
    } else if (videoLink.includes('video/')) {
      const vidId = videoLink.split('video/')[1].split('/')[0];
      launchWorker(vidId);
    } else if (videoLink.includes('u.be/')) {
      const vidId = videoLink.split('be/')[1].split('?')[0];
      launchWorker(vidId);
    } else if (videoLink === 'test') {
      props.changeStreamProperties({
        ownerId: '',
        thumbnailUrl: 'https://i.ytimg.com/vi/HwsGz6csNA0/maxresdefault.jpg',
        title: '',
        videoId: 'test',
      });
    } else {
      setError('invalidUrl');
    }
  };

  useEffect(() => {
    const id = sessionStorage.getItem('gv-videoId');
    const storedTitle = sessionStorage.getItem('gv-title');
    const storedThumbnail = sessionStorage.getItem('gv-thumbnailUrl');
    const storedOwnerId = sessionStorage.getItem('gv-ownerId');
    if (
      id !== null &&
      storedTitle !== null &&
      storedThumbnail !== null &&
      storedOwnerId !== null
    ) {
      const streamProps = {
        ownerId: storedOwnerId,
        thumbnailUrl: storedThumbnail,
        title: storedTitle,
        videoId: id,
      };
      props.changeStreamProperties(streamProps);
    }
  }, []);

  if (props.streamInfo.videoId === '' || props.ban !== null) {
    return (
      <WelcomeDialog
        passVideo={receiveVideo}
        ban={props.ban}
        error={error}
        variant={0}
      />
    );
  }
  return (
    <div>
      <FormattedMessage {...messages.pageTitle}>
        {(title) => (
          <Helmet>
            <title>{title}</title>
          </Helmet>
        )}
      </FormattedMessage>
      <TopBar>
        <StreamInfo>
          <StreamImg alt="Thumbnail" src={props.streamInfo.thumbnailUrl} />
          <StreamTitle>{props.streamInfo.title}</StreamTitle>
          <StyledButton onClick={leaveStream}>
            <FormattedMessage {...messages.leaveStreamBtn} />
          </StyledButton>
        </StreamInfo>
        <TopButtons>
          <HistoryWidget />
          <SupportInformation />
          <SettingsDialog />
        </TopButtons>
      </TopBar>
      <YoutubeWorker videoId={props.streamInfo.videoId} apiKey={API_KEY} />
    </div>
  );
};

GiveawayPage.propTypes = {
  ban: PropTypes.object,
  changeStreamProperties: PropTypes.func.isRequired,
  sendTelemetryData: PropTypes.func,
  streamInfo: PropTypes.object.isRequired,
};

const mapStateToProps = createStructuredSelector({
  ban: makeSelectBanStatus(),
  streamInfo: makeSelectStreamInfo(),
});

export function mapDispatchToProps(dispatch) {
  return {
    changeStreamProperties: (st) => dispatch(changeStreamProperties(st)),
    sendTelemetryData: (st) => dispatch(sendTelemetryData(st)),
    dispatch,
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(GiveawayPage);
