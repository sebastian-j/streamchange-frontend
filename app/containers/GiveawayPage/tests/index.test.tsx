import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';

import GiveawayPage from '../index';
import configureStore from '../../../configureStore';


describe('<GiveawayPage />', () => {
  let store;
  beforeAll(() => {
    store = configureStore({
      authKey: '',
      ban: null,
      stream: {
        ownerId: '',
        thumbnailUrl: '',
        title: '',
        videoId: '',
      },
    });
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <GiveawayPage />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
