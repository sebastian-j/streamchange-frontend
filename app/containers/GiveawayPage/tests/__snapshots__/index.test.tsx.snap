// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiveawayPage /> should render and match the snapshot 1`] = `
<div
  class="sc-jrQzAO kRluvY"
>
  <span
    class="sc-hBUSln gasaX"
  >
    Welcome to StreamChange
  </span>
  <div>
    <span
      class="sc-hBUSln gasaX"
    >
      Select theme
    </span>
    <label
      for="themeLight"
    >
      <input
        checked=""
        class="sc-egiyK dHVmLx"
        id="themeLight"
        name="theme"
        type="radio"
        value="0"
      />
      <div
        class="sc-kDTinF iBZAcK box left"
      >
        <img
          alt="light"
          src="IMAGE_MOCK"
        />
        <span
          class="sc-ksdxgE dwWHTf"
        >
          Light
        </span>
      </div>
    </label>
    <label
      for="themeDark"
    >
      <input
        class="sc-egiyK dHVmLx"
        id="themeDark"
        name="theme"
        type="radio"
        value="1"
      />
      <div
        class="sc-kDTinF iBZAcK box right"
      >
        <img
          alt="dark"
          src="IMAGE_MOCK"
        />
        <span
          class="sc-ksdxgE dwWHTf"
        >
          Dark
        </span>
      </div>
    </label>
  </div>
  <div>
    <span
      class="sc-hBUSln gasaX"
    >
      Select language
    </span>
    <label
      for="pl"
    >
      <input
        class="sc-egiyK dHVmLx"
        id="pl"
        name="language"
        type="radio"
        value="pl"
      />
      <div
        class="sc-iqseJM hcrhBp box left"
      >
        <span
          class="sc-crHmcD etoWxl"
        >
          <img
            alt="pl"
            src="IMAGE_MOCK"
          />
          pl
        </span>
      </div>
    </label>
    <label
      for="en"
    >
      <input
        checked=""
        class="sc-egiyK dHVmLx"
        id="en"
        name="language"
        type="radio"
        value="en"
      />
      <div
        class="sc-iqseJM hcrhBp box right"
      >
        <span
          class="sc-crHmcD etoWxl"
        >
          <img
            alt="en"
            src="IMAGE_MOCK"
          />
          en
        </span>
      </div>
    </label>
  </div>
  <div>
    <button
      class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root sc-bqiRlB gNEjRC css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <span>
        Save
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="sc-furwcr dCipSX"
  >
    <img
      alt="cookie"
      class="sc-iCfMLu gLHhKf"
      src="IMAGE_MOCK"
    />
    <span
      class="sc-gKclnd dnVwcj"
    >
      This site uses only necessary functional cookies. There is no tracking or personalized ads based on cookies.
    </span>
    <button
      class="sc-jRQBWg hvQnvK"
      type="button"
    >
      Accept
    </button>
  </div>
</div>
`;
